### ISSUE TITLE
Alt + Arrow Inserting Symbols on Alt Release (Windows)

### ISSUE THREAD

**Issue by tkefauver** [TolikPylypchuk/SharpHook#67]
This wasn't an issue before version 5 but now when using <kbd>alt</kbd><kbd>up</kbd> or <kbd>alt</kbd><kbd>down</kbd> in a text editor (shortcuts to shift line of text up/down) a random character is inserted into the text after releasing <kbd>alt</kbd>.  

Repeatedly pressing <kbd>up</kbd> or <kbd>down</kbd> while holding <kbd>alt</kbd> changes the random character thats inserted and from researching it, this is intended behavior when using numpad and holding <kbd>alt</kbd> but **not** the arrow keys. I only have a laptop keyboard that doesn't have a numpad so I can't test what's going on much but here's a [link](https://www.rapidtables.com/code/text/alt-codes.html) with more info about it.

All I can think of that would cause this is inside libuiohook somehow the arrow keys are being mapped to numpad arrows or something unusual because <kbd>alt</kbd> is down. I really don't know but have confirmed it wasn't an issue before version 5. 

* I'm using the async `SimpleGlobalHook`

**Comment by TolikPylypchuk** [comment]
Hi! Thanks for opening this issue! Could you please share some logs, e.g. like this?

```c#
hook.KeyPressed += (sender, e) => Console.WriteLine(e.RawEvent);
hook.KeyReleased += (sender, e) => Console.WriteLine(e.RawEvent);
```

It would be great to see some [libuiohook debug logs](https://sharphook.tolik.io/v5.1.1/articles/logging.html) as well.

Also, could you please tell me which OS you're using?

**Comment by tkefauver** [comment]
Yeah no problem!
My OS is Windows 11 22H2 22621.2428

The logs are below (key press/releases are mixed w/ the libuiohook logs). 

- Just to explain what I was doing: its pressing <kbd>alt</kbd> with<kbd>up</kbd> and <kbd>down</kbd> a few times.
- Also all the '￿' KeyChars  show up as 'xBFxEFxEF' symbols in notepad++ not sure what those mean though.

```
hook_set_dispatch_proc [41]: Setting new dispatch callback to 0X00007FF85DC900E4.
hook_run [349]: SetWindowsHookEx() successful.
refresh_locale_list [369]: GetKeyboardLayoutList(0, NULL) found 1 layouts.
refresh_locale_list [384]: Received 1 locales.
get_keyboard_layout_file [325]: Found keyboard layout "00000409".
refresh_locale_list [458]: Loading layout for 0X0000000004090409: KBDUS.DLL.
load_input_helper [623]: refresh_locale_list() found 1 locale(s).
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768439228; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768439729; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768439901; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768440057; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768440229; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768440386; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768440559; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768440731; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768440887; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0X27 pressed. (0X27)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768441059; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcRight; RawCode = 39; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X27 released. (0X27)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768441216; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcRight; RawCode = 39; KeyChar = ￿
dispatch_key_release [237]: Key 0XA012 released. (0XA4)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768441544; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
dispatch_key_press [165]: Key 0X27 pressed. (0X27)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768442623; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcRight; RawCode = 39; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X27 released. (0X27)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768442826; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcRight; RawCode = 39; KeyChar = ￿
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768445516; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0X28 pressed. (0X28)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768445704; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcDown; RawCode = 40; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X28 released. (0X28)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768445924; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcDown; RawCode = 40; KeyChar = ￿
dispatch_key_release [237]: Key 0XA012 released. (0XA4)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768448050; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
dispatch_key_press [165]: Key 0X28 pressed. (0X28)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768451272; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcDown; RawCode = 40; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X28 released. (0X28)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768451491; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcDown; RawCode = 40; KeyChar = ￿
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768452805; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0X26 pressed. (0X26)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768453133; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcUp; RawCode = 38; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X26 released. (0X26)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768453321; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcUp; RawCode = 38; KeyChar = ￿
dispatch_key_release [237]: Key 0XA012 released. (0XA4)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768454416; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
dispatch_key_press [165]: Key 0X28 pressed. (0X28)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768456542; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcDown; RawCode = 40; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X28 released. (0X28)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768456699; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcDown; RawCode = 40; KeyChar = ￿
dispatch_key_press [165]: Key 0X28 pressed. (0X28)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768462047; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcDown; RawCode = 40; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X28 released. (0X28)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768462250; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcDown; RawCode = 40; KeyChar = ￿
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768463689; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768464173; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768464345; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768464533; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768464689; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768464861; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768465065; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0XA012 pressed. (0XA4)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768465237; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_press [165]: Key 0X26 pressed. (0X26)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768465393; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcUp; RawCode = 38; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X26 released. (0X26)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768465565; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcUp; RawCode = 38; KeyChar = ￿
dispatch_key_press [165]: Key 0X26 pressed. (0X26)
dispatch_event [69]: Dispatching event type 4.
UioHookEvent: Type = KeyPressed; Time = 1698768465659; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcUp; RawCode = 38; KeyChar = ￿
keycode_to_unicode [590]: Using keyboard layout 0X0000000004090409.
dispatch_key_release [237]: Key 0X26 released. (0X26)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768465816; Mask = LeftAlt; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcUp; RawCode = 38; KeyChar = ￿
dispatch_key_release [237]: Key 0XA012 released. (0XA4)
dispatch_event [69]: Dispatching event type 5.
UioHookEvent: Type = KeyReleased; Time = 1698768465909; Mask = None; Reserved = None; Keyboard = KeyboardEventData: KeyCode = VcLeftAlt; RawCode = 164; KeyChar = ￿
```

**Comment by TolikPylypchuk** [comment]
Thanks, I will look into it in the next couple days.

**Comment by TolikPylypchuk** [comment]
Hi @tkefauver, sorry for the long delay, but I've only been able to look into this issue today. And it can indeed be reproduced reliably. Hopefully, I'll be able to fix it soon.

**Comment by TolikPylypchuk** [comment]
I've found the code line in libuiohook which causes this issue, but I don't yet know why it does so and how to fix that. I'm going to share it here for visibility, and maybe someone would be able to help if I don't find the solution in the nearest future.

In the key pressed handler, there's the following code which determines if a key press should cause a character to be typed (and whether to raise the key typed event):

```c
BYTE keyboard_state[256] = { 0 };
GetKeyState(0); // This apparently forces GetKeyboardState to get more up-to-date data.

// Get current keyboard state (to know which modifiers have been pressed, as well as Caps Lock).
BOOL success = GetKeyboardState(keyboard_state);

if (!success) {
    logger(LOG_LEVEL_ERROR, "%s [%u]: GetKeyboardState() failed! (%#lX)\n",
        __FUNCTION__, __LINE__, (unsigned long)GetLastError());
    return 0;
}

// Look up the Unicode code for the key without changing the keyboard state.
int result = ToUnicodeEx(keycode, scancode, keyboard_state, buffer, size, 1 << 2, locale_current->id);
```

The `ToUnicodeEx` function is used to determine the character, but it needs the current state of the keyboard to do that correctly (to know, e.g. whether the `Shift` is pressed). `GetKeyboardState` does exactly that, but for some reason, it doesn't return up-to-date data, so `ToUnicodeEx` doesn't work correctly. Somehow, calling `GetKeyState(0)` forces `GetKeyboardState` to return up-to-date data instead. This is where it gets really confusing - calling `GetKeyState` is what also makes editors like Visual Studio or Notepad++ to insert a character on `Alt+Up` or `Alt+Down`. If this function call is removed, then this behavior is not present anymore, but in doing so, the key typed events now contain incorrect characters.

The libuiohook code used in SharpHook before 5.0.0 didn't use `ToUnicodeEx` - it determined (or rather tried to determine) the characters by other means which were really obscure and didn't work as well as `ToUnicodeEx`.

I'm going to investigate further what happens here and why.

**Comment by tkefauver** [comment]
Definitely sounds confusing! 

One thing that stood out to me you probably already realize is the false KeyTyped event only happens when `Alt` is released.

**Comment by TolikPylypchuk** [comment]
Yeah, it somehow messes up the keyboard state, I think, but I'm not sure on that one yet.

**Comment by TolikPylypchuk** [comment]
Ok, so I've figured out that the culprit of this issue was the `ToUnicodeEx` function itself, and not `GetKeyState` as I've previously thought. Calling `GetKeyState` is what makes `ToUnicodeEx` use the latest keyboard state, and it then proceeds to mess it up.

`ToUnicodeEx` is really weird about changing the keyboard state. One of the parameters to this function is the flags - previously I've set bit 2 since according to the docs:

> If bit 2 is set, keyboard state is not changed (Windows 10, version 1607 and newer)

But it looks like it changes the keyboard state somehow after all. There is another flag that I've found:

> If bit 0 is set, a menu is active. In this mode **Alt+Numeric keypad** key combinations are not handled.

I've tried it out with this bit set as well, and behold, it worked! Editors no longer insert characters on `Alt+Arrow`. I hope it doesn't introduce some other weird bugs, but those will probably still be better than changing the keyboard state on every key press.

I will release a new version with the fix shortly.

**Comment by TolikPylypchuk** [comment]
Version 5.1.2 which fixes this problem has been released, so this issue can be closed.

**Comment by TolikPylypchuk** [comment]
Hi @tkefauver! Not sure if this is relevant for you, but the upcoming version 6.0 of SharpHook will include a revert of this issue's fix. Turns out that adding the Alt flag to `ToUnicodeEx` [breaks Alt codes](https://github.com/TolikPylypchuk/SharpHook/issues/125).

The good news is that the issue you've described - editors inserting characters on <kbd>Alt</kbd>+<kbd>Arrow</kbd> - cannot be reproduced anymore on Windows 11. The bad news is that it still can be reproduced on Windows 10. Reverting this fix will be the lesser of two evils as I think having working Alt codes is more important that editors inserting characters on <kbd>Alt</kbd>+<kbd>Arrow</kbd> on Windows 10. I'm not going to spend time trying to fix this at the moment since this is a Windows 10-specific issue but may change my mind in the future.

**Comment by TolikPylypchuk** [comment]
Hi again @tkefauver! Unfortunately, the issue can also be reproduced on Windows 11, but I was only able to reproduce it in Visual Studio - other editors like Visual Studio Code work fine.

I've just released version 6.2.0 which lets the application disable events of type `KeyTyped` which removes this issue altogether. It is recommended to disable them if the application doesn't use `KeyTyped`:

```c#
UioHookProvider.Instance.KeyTypedEnabled = false;
```
